### PR TITLE
More advices around search speed and disk usage.

### PR DESCRIPTION
**Pull request by jpountz**
It adds notes about:
 - how preference can help optimize cache usage
 - the fact that too many replicas can hurt search performance due to lower
   utilization of the filesystem cache
 - how index sorting can improve _source compression
 - how always putting fields in the same order in documents can improve _source
   compression

### CODE REVIEW

File: docs/reference/how-to/disk-usage.asciidoc [hunk 1]
```diff
@@ -158,3 +158,27 @@ on disk usage. In particular, integers should be stored using an integer type
 stored in a `scaled_float` if appropriate or in the smallest type that fits the
 use-case: using `float` over `double`, or `half_float` over `float` will help
 save storage.
+
+[float]
+=== Use index sorting to colocate similar documents
+
+Elasticsearch compresses multiple documents at once in order to improve the
```
Comment (nik9000) [inline docs/reference/how-to/disk-usage.asciidoc]: Maybe something like "When Elasticsearch stores `_source`, it compresses multiple documents at once to improve the overall compression ratio...." so we don't get people thinking that doc values and the inverted index bits are stored like this.

File: docs/reference/how-to/disk-usage.asciidoc [hunk 2]
```diff
@@ -158,3 +158,27 @@ on disk usage. In particular, integers should be stored using an integer type
...
+overall compression ratio. For instance it is very common that documents share
+the same field names, and quite common that they share some field values,
+especially on fields that have a low cardinality or a zipfian distribution.
```
Comment (nik9000) [inline docs/reference/how-to/disk-usage.asciidoc]: zipfian might deserve a link to wikipedia.

File: docs/reference/how-to/disk-usage.asciidoc [hunk 3]
```diff
@@ -158,3 +158,27 @@ on disk usage. In particular, integers should be stored using an integer type
...
+
+Documents that are compressed together are documents that are colocated in the
```
Comment (nik9000) [inline docs/reference/how-to/disk-usage.asciidoc]: I'd twist this around to something like "By default documents are compressed together in the order that they are added to the index. If you enabled index sorting then instead they are compressed in sorted order. Sorting documents with similar structure, fields, and values together *should* improve the compression ratio." Or something like that. It feels more active that way. I dunno.

File: docs/reference/how-to/search-speed.asciidoc [hunk 4]
```diff
@@ -326,3 +326,44 @@ queries, they should be mapped as a `keyword`.
 <<index-modules-index-sorting,Index sorting>> can be useful in order to make
 conjunctions faster at the cost of slightly slower indexing. Read more about it
 in the <<index-modules-index-sorting-conjunctions,index sorting documentation>>.
+
+[float]
+=== Use `preference` to optimize cache utilization
+
+There are multiple caches that can help with search performance, such as the
+filesystem cache, the <<shard-request-cache,request cache>> or the
```
Comment (nik9000) [inline docs/reference/how-to/search-speed.asciidoc]: Maybe make filesystem cache a link to https://en.wikipedia.org/wiki/Page_cache ?

Comment (nik9000) [inline docs/reference/how-to/search-speed.asciidoc]: s/or/and/

File: docs/reference/how-to/search-speed.asciidoc [hunk 5]
```diff
@@ -326,3 +326,44 @@ queries, they should be mapped as a `keyword`.
...
+<<query-cache,query cache>>. Yet all these caches are maintained at the node
+level, meaning that if you run the same request twice in a row, have 1
+<<glossary-replica-shard,replica>> or more and use the default routing
```
Comment (nik9000) [inline docs/reference/how-to/search-speed.asciidoc]: I'd s/use the default routing algorithm, which is round-robin,/use round-robin, the default routing algorithm/

**Review by nik9000 — APPROVED** [review]
Left a few minor things but LGTM.

**Review by martijnvg — APPROVED** [review]
LGTM

**Pull request merged**